odxtools/decodestate.py

```diff
@@ -87,32 +93,125 @@ def extract_atomic_value(
+            else:
+                odxassert(
+                    base_type_encoding in (None, Encoding.NONE),
+                    f"Specified illegal encoding {base_type_encoding} for string object")
+
+                # if no encoding has been specified explicitly, we
+                # make assumptions by looking at the data type
+                if base_data_type == DataType.A_UTF8STRING:
+                    internal_value = raw_value.decode("utf-8", errors=text_errors)
+                elif base_data_type == DataType.A_UNICODE2STRING:
+                    text_encoding = "utf-16-be" if is_highlow_byte_order else "utf-16-le"
+                    internal_value = raw_value.decode(text_encoding, errors=text_errors)
+                else:
+                    odxassert(base_data_type == DataType.A_ASCIISTRING)
+                    # The spec says ASCII, meaning only character
+                    # values 0-127.  In practice, vendors use
+                    # iso-8859-1, aka latin-1, because iso-8859-1
+                    # never fails since it has a valid character
+                    # mapping for every possible value
+                    internal_value = raw_value.decode("iso-8859-1", errors=text_errors)
+
+        # ... integers, ...
+        elif base_data_type in (DataType.A_INT32, DataType.A_UINT32):
+            if not isinstance(raw_value, int):
+                odxraise(f"Raw value must be of integer type, not {type(raw_value).__name__}")
+
+            # BCD encodings
+            if base_type_encoding == Encoding.BCD_P:
+                # packed BCD
+                tmp2 = raw_value
+                internal_value = 0
+                factor = 1
+                while tmp2 > 0:
+                    internal_value += (tmp2 & 0xf) * factor
+                    factor *= 10
+                    tmp2 >>= 4
+            elif base_type_encoding == Encoding.BCD_UP:
+                # unpacked BCD
+                tmp2 = raw_value
+                internal_value = 0
+                factor = 1
+                while tmp2 > 0:
+                    internal_value += (tmp2 & 0xf) * factor
+                    factor *= 10
+                    tmp2 >>= 8
+            elif base_type_encoding == Encoding.ONEC:
+                # one-complement
+                sign_bit = 1 << (bit_length - 1)
+                if raw_value < sign_bit:
+                    internal_value = raw_value
+                else:
+                    # python defines the bitwise inversion of a
+                    # positive integer value x as ~x = -(x + 1).
+                    internal_value = -((1 << bit_length) - raw_value - 1)
+            elif base_type_encoding == Encoding.TWOC:
+                # two-complement
+                sign_bit = 1 << (bit_length - 1)
+                if raw_value < sign_bit:
+                    internal_value = raw_value
+                else:
+                    internal_value = -((1 << bit_length) - raw_value)
```

Comment:
There is a special case of 2C you need to check, if bit length = 1, and 2C is used, then the encoding should map 0 to 0, and 1 to -1

Reply:
I think this already is what happens here:

```
raw_value = 1
-> - (2 -1) = -1
raw value  = 0
-> 0
```